README.md

## `Star`是什么?

`Star`是 STock Analysis and Research tool 的简称，主要用于A股股票追踪分析。`star`目前的主要作用是根据您所设定的股票目标价及当前价计算出相对该**目标价**的上涨空间，即`(目标价-当前价)/当前价x100%`，然后将此上涨空间按顺序排列方便您从中找到上涨空间最大的股票作为买入参考。`star`不提供股价预测和股票交易服务，目前只是单纯的计算，所以它的价值很大程度上取决于您所设定的目标价的有效性，目标价可以选择近期股票的最高价格或者分析师给出的价格，在大牛市里面这样的目标价还是比较容易达到的。同时`star`还具有股票基本报价信息查询、看盘、股票董监高持股信息批量查询、查阅未来30日财经日历等功能，未来可能会加入更多功能，也欢迎大家[反馈建议](https://github.com/hustcer/star/issues/new)。


## Donate

这个工具的开发和维护花费了我大量的时间和精力，如果`Star`对你有所帮助，也欢迎您给予支持，谢谢：
![赞助Star](https://uzfin.com/static/images/backing.jpg "赞助Star")


## 安装及使用

`Star`是基于`node.js`的所以先要安装`node.js`, 推荐您使用最新版本的`node.js`，同时最好使用 Mac/Linux系统，Windows下目前还没测试过，理论上可以使用，如果具体使用过程中遇到问题也可以[在此反馈](https://github.com/hustcer/star/issues/new)；具体安装步骤如下：

1. 安装 [node.js](https://nodejs.org/), version >= 4.2.0，始终建议您采用最新版本的 `node.js`;
1. 安装`star`: `sudo npm i -g star`，注意一定记得加上 `-g` 参数，这样在任意路径下都可以调用`star`命令，否则可能找不到可执行文件;
1. 运行`star`: 直接执行 `star` 即可，第一次运行的时候需要加`-f`或`--file`参数指定股票数据文件(推荐采用绝对路径,详见后文说明)，之后可以省略该参数;
1. 更多帮助可以参考: `star --help`;

注意：如果`npm`安装很慢(由于一些众所周知的原因)，可以使用[`cnpm`](https://npm.taobao.org/)代替，下同；

### 从源码安装

如需从源码安装用如下步骤替换前面的**第二步**即可：

1. 在安装 `git` 之后 clone `star`源码：`git clone git@github.com:hustcer/star.git`;
1. 在`star`目录下安装依赖的 `node` 模块: `cd star && sudo npm install`或者`sudo npm install --production`，加上`--production`参数则不安装开发依赖的相关模块比如`gulp`等;
1. 然后执行 `./star.js` 即可，推荐创建软链接：`ln -s /absolute/path/to/star.js /usr/local/bin/star`, 这样就可以在任意位置通过`star`命令进行调用；

## 升级

本工具会不定时更新，升级方法如下：

1. 【可选】升级node.js, 建议您使用[`nvm`](https://github.com/creationix/nvm)来管理node这样升级起来很方便，mac下也可以使用[`brew`](http://brew.sh/)来安装、升级node；
2. 更新 `star`：`sudo npm update -g star`;

## FAQ

搜集了一些常见问题及解决办法，参考[这里](FAQ.md)

## What' New ?

变更详情 [**见此**](history.md)，主要变更如下：

- 【v0.5.0】 修复财经日历获取异常；修复深交所股票董监高增减持记录查询异常.
- 【v0.3.9】 修复财经日历内容解码问题.
- 【v0.3.8】 修复 node.js V6+ 环境下帮助文档输出被截断的 bug.
- 【v0.3.5】 优化命令行中数据表格展示；
- 【v0.3.2】 股票追踪增加对融资融券标的的过滤支持，可以通过`-M`或者`--margin`实现；
- 【v0.3.1】 修复深市股票董监高持股变动查询时的分页显示问题；
- 【v0.3.1】 `-g`或`--grep`参数支持对股票代码的过滤；
- 【v0.3.0】 通过`-i`或`--insider`参数查询董监高持股变化信息的时候也可以不指定股票代码，此时查询沪深两市最近的董监高交易记录，按交易时间倒序排列；
- 【v0.3.0】 未指定股票代码查询董监高持股变动信息时允许通过`--market`参数来指定板块儿：SZM-深圳主板, SZGEM-深圳创业板, SZSME-深圳中小板, SHM-上海主板. 大小写不敏感; 指定股票代码时会根据股票代码来判断属于哪个市场，所以`--market`参数无效。
- 【v0.3.0】 可以通过`-i --top-buy`参数查询最近买入总额排名前列的股票，`--span`参数代表查询时间区间：1m~12m，即最近一个月到最近12个月；数据源: [uzfin.com](http://uzfin.com)
- 【v0.3.0】 可以通过`-i --top-sell`参数查询最近卖出总额排名前列的股票，`--span`参数代表查询时间区间：1m~12m，即最近一个月到最近12个月；数据源: [uzfin.com](http://uzfin.com)
- 【v0.3.0】 修复查询深市个股董监高持股变动信息时的分页问题，可以通过`-p`或者`--page`参数指定要查询的分页，默认为第一页。
- 【v0.2.8】 增加`--remove`参数筛选证券名称和备注里面**不包含**指定关键词的股票，多个关键词之间用','或'，'隔开.
- 【v0.2.8】 股票追踪增加按`bdiff`即根据`(s.price - s.cheap)/s.price`指标排序.
- 【v0.2.8】 股票追踪增加按`sdiff`即根据`(s.price - s.expensive)/s.price`指标排序.
- 【v0.2.7】 增加`--lteb [pct]` 参数用于筛选出使得如下条件成立的股票：`100*(s.price - s.cheap)/s.price <= pct`，如果`pct`为空则上述条件退化为：`s.price <= s.cheap`.
- 【v0.2.7】 增加`--gtes [pct]` 参数用于筛选出使得如下条件成立的股票：`100*(s.price - s.expensive)/s.price >= pct`，如果`pct`为空则上述条件退化为：`s.price >= s.expensive`.
- 【v0.2.6】 新增`--lteb` 和 `--gtes`参数，用于筛选当前价小于等于适合买入的便宜价格或者大于等于应该卖出的昂贵价格的股票(分别对应股票配置里面的`cheap`和`expensive`价格)。
- 【v0.2.6】 允许通过 `star -wo` 或者 `star -w -o` watch 当前持有的股票, 原有的 fallback 机制不变。
- 【v0.2.5】 改进内部交易查询等在命令行终端输出时的对齐问题。
- 【v0.2.5】 看盘股票列表的数目从 **20** 调整到 **25**，批量请求的数目也作相应地调整。
- 【v0.2.4】 财经日历的事件列表输出排版改进——去掉冗余的空白字符。
- 【v0.2.4】 股票配置文件的`symbols`部分添加重复股票检查功能。


## 具体功能列表

### 股票追踪

- 设置股票的买入价、卖出价、目标价、星级、备注等，可以自动获取股票的当前价格并且计算距离目标价的上涨空间：`(目标价-当前价)/当前价x100%`，上涨空间最大的股票通常也是最有利可图的，这也是本工具的最主要功能，方便快速决定值得买入的股票；
- 支持两个股票信息获取数据源：腾讯和新浪股票数据源，万一其中一个有问题可以通过`-d`或`--data`参数切换到另一个；
- 可以按照股票的code(证券代码)、star(星级)、price(当前价)、incp(当前涨幅)、bdiff( `(s.price - s.cheap)/s.price` )、sdiff( `(s.price - s.expensive)/s.price` )、targetp(上涨空间: `(s.target - s.price)/s.price` )、capacity(市值)、pe(PE)、pb(PB)等条件进行升、降序排序，其中后三项排序只有在采用腾讯数据接口的时候才支持；
- 股票数据比较多的时候可以设置每次显示多少条数据(通过`-l`或`--limit`参数)，并且进行分页(通过`-p`或`--page`参数)；
- 可以设置是否关注、持有某股票，并根据这些条件进行过滤：可以显示所有股票(-a, --all)、只显示持有的股票(-o, --hold)、只显示不再关注的股票(-I, --ignore)；
- 可以通过`-e`或`--exclude`参数排除所有证券代码以300,600,002或者000开头的股票，多个前缀之间以','或者'，'分隔；
- 可以通过`-c`或`--contain`参数只显示所有证券代码以300,600,002或者000开头的股票，多个前缀之间以','或者'，'分隔；
- 可以通过`-f`或`--file`参数指定股票文件路径，并自动保存该路径，下次执行命令的时候不必重复输入；
- 可以通过`-g`或`--grep`参数对股票列表里的股票名、股票代码、备注字段进行搜索、过滤，多个关键词之间以','或者'，'分隔；
- 可以通过`--remove`参数筛选证券名称和备注里面**不包含**指定关键词的股票，多个关键词之间用','或'，'分隔.
- 可以通过`-L`或`--lte`参数过滤出当前价到目标价的上涨空间百分比小于等于指定百分比的股票；
- 可以通过`-G`或`--gte`参数过滤出当前价到目标价的上涨空间百分比大于等于指定百分比的股票；
- 可以通过`-U`或`--under`参数过滤出股票星级等于或在指定星级之下的股票；
- 可以通过`-A`或`--above`参数过滤出股票星级等于或在指定星级之上的股票；
- 可以通过`-M`或`--margin`参数过滤出支持融资融券的股票；
- 可以通过`--lteb [pct]`参数用于筛选出使得如下条件成立的股票：`100*(s.price - s.cheap)/s.price <= pct`，如果`pct`为空则上述条件退化为：`s.price <= s.cheap`，其中`s.price`为股票当前价格，之所以使用`s.price`作为分母而不是`s.cheap`主要是为了减少因为后者的随意性而带来的偏差；
- 可以通过`--gtes [pct]`参数用于筛选出使得如下条件成立的股票：`100*(s.price - s.expensive)/s.price >= pct`，如果`pct`为空则上述条件退化为：`s.price >= s.expensive`，其中`s.price`为股票当前价格，之所以使用`s.price`作为分母而不是`s.expensive`主要是为了减少因为后者的随意性而带来的偏差；


### 股票报价查询

- 通过证券代码查询相应股票的基本报价信息，多个代码之间以','或者'，'分隔，一次最多查询25个，例如：`star 300315,600048，600015`；

### 股票报价自动更新/看盘

- 可以通过`-w`或`--watch`参数看盘，在该模式下股票报价数据会自动更新，目前更新时间间隔为3.6秒，例如：`star -w 601179,600118,600893,000712,002625`，最终执行效果类似于在终端执行`top`命令；
- 每次输入这些股票代码可能比较麻烦，所以为了方便起见可以修改股票配置文件的 `watchList` 部分内容，详见后文说明，这样每次只需要执行`star -w` 命令即可；
- 如果 `watchList` 部分内容为空则自动会查找 `symbols` 部分股票里面 `hold` 为 `true` 的部分股票；
- 如果在使用`-w`或`--watch`参数的同时再加上`-o`或`--hold`参数则忽略`watchList`里面的股票直接加载`symbols` 部分股票里面 `hold` 为 `true` 的部分股票，比如`star -w -o`或者`star -wo`；
- 需要说明的是，目前看盘时股票数目有限制，最多同时观察25只股票，这个数目在大多数情况下已经足够了，加上这个限制主要是为了减少数据请求的次数；

### 董监高持股变动信息批量查询

- 可以通过`-i`或`--insider`参数加上对应的证券代码，查询相应股票的高管交易记录，多个代码之间以','或者'，'分隔，支持对深市主板、中小板、创业板及沪市股票的董监高持股变动信息进行查询；
- 通过`-i`或`--insider`参数查询董监高持股变化信息的时候也可以不指定股票代码，此时查询沪深两市最近董监高交易记录，按交易时间倒序排列，可以通过`-p`或`--page`参数进行翻页；
- 可以通过`--from 2014/01/01 --to 2015/07/09`参数限定高管交易记录的查询起始时间，默认从当前时间开始往前推一年；
- 可以通过`--span 6m`参数限定高管交易记录的查询时间跨度月数，取值范围从`1m`到`24m`，`6m`表示从现在开始往前推6个月，默认为`12m`，例如：`star -i 000768,002456,600118,300036 --span 3m`；
- 需要注意的是目前一次最多可以查询25只股票，如果你有更多股票可以分多次进行查询，加上这个限制同样是为了减少数据请求次数和服务器压力；
- 未指定股票代码查询董监高持股变动信息时允许通过`--market`参数来指定板块：SZM-深圳主板, SZGEM-深圳创业板, SZSME-深圳中小板, SHM-上海主板. 大小写不敏感。**注意：当指定股票代码的时候会自动忽略该参数**；
- 可以通过`-i --top-buy`参数查询最近买入总额排名前列的股票，`--span`参数代表查询时间区间：1m~12m，即最近一个月到最近12个月；
- 可以通过`-i --top-sell`参数查询最近卖出总额排名前列的股票，`--span`参数代表查询时间区间：1m~12m，即最近一个月到最近12个月；

### 未来30日财经日历

- 可以通过`-C`或`--cal`参数查询中国股市未来30日题材前瞻；

### 其它

- 还有更多功能正在酝酿之中，同时也欢迎大家广泛提建议；

## 股票数据文件说明

`Star`的运行需要一个股票数据文件，如果未指定股票数据文件默认情况下会使用当前目录下的 [`symbols.yaml`](https://github.com/hustcer/star/blob/master/symbols.yaml) 文件，该文件采用`yaml`语法编写，修改的时候尤其要注意缩进问题，否则可能会解析错误。一个典型的股票数据文件格式说明如下：

### 股票追踪数据说明

```yaml
symbols:
  -
    name      : '掌趣科技'            # <字符串>公司名称
    code      : '300315'             # <字符串>证券代码，必填，不可错
    cheap     : 16                   # <数字>买点价位，在此价格附近买入有利可图
    expensive : 22                   # <数字>卖点价位，在此价格附近可以卖出获利了结
    target    : 23                   # <数字>目标价位
    star      : 3                    # <数字>股票评级，越高越值得购买，范围：1~5
    hold      : false                # <true/false> true 表示当前持有该股票,反之不持有
    watch     : true                 # <true/false>是否关注此股票，如果 true 则关注，否则忽略
    comment   : '手游;高送转;自设目标'  # <字符串>行业/概念/题材等备注,后期可用于搜索过滤
  -
    name      : '启明星辰'
    code      : '002439'
    cheap     : 50
    expensive : 88
    target    : 78
    star      : 3
    hold      : false
    watch     : true
    comment   : '计算机;信息安全;腾讯合作;分析师推荐'
```

### 股票报价自动更新/看盘数据格式说明

```yaml
# 通过`star.js -w`或`star.js --watch`看盘时如果没有传入证券代码参数则从这里取得看盘股票列表
# 如果 watchList 列表为空或者不存在则从 symbols 列表中获取 hold 为 true 的股票
watchList:
  -
    name :  '安信信托'
    code :  '600816'
  -
    name :  '中航飞机'
    code :  '000768'
  -
    name :  '华胜天成'
    code :  '600410'
  -
    name :  '中国卫星'
    code :  '600118'
  -
    name :  '中航动力'
    code :  '600893'
  -
    name :  '锦龙股份'
    code :  '000712'
```

## 运行截图

1. 帮助文档
![](https://github.com/hustcer/star/blob/master/snapshot/help.png)

2. 股票筛选结果
![](https://github.com/hustcer/star/blob/master/snapshot/snapshot.png)

3. 股票基本报价信息查询
![](https://github.com/hustcer/star/blob/master/snapshot/query.png)

4. 股票看盘(报价数据自动更新)
![](https://github.com/hustcer/star/blob/master/snapshot/watch.png)

5. 批量查询股票董监高增减持信息
![](https://github.com/hustcer/star/blob/master/snapshot/insider.png)

6. 董监高增减持排行榜
![](https://github.com/hustcer/star/blob/master/snapshot/topBuy.png)

7. 中国股市未来30日题材前瞻
![](https://github.com/hustcer/star/blob/master/snapshot/cal.png)


history.md

## Change History

### v0.5.0

- Update node modules.
- fix: Fix SZ market insider trading query.
- fix: Fix finance calendar query.

### v0.3.9

- Update node modules: async, blessed-contrib, js-yaml, lodash and eslint.
- Update node modules: async, bluebird, moment, eslint and blessed-contrib.
- Update node modules: eslint,eslint-plugin-import,eslint-config-airbnb,js-yaml,lodash,moment and request.
- Update node modules:async,blessed-contrib,lodash,moment and request.
- Unescape html entities and collapse whitespace of fin cal by using strman.
- Update eslint and fix eslint check.
- Update node.js modules:async, eslint and eslint-plugin-import.
- Update node modules:bluebird, iconv and strman.
- Update node modules: strman, eslint, eslint-plugin-import and gulp-eslint.
- Add `sloc` counter task for gulp.
- Use numbro instead of numeral.
- Update node modules: moment, request and strman.
- Update node modules: async, eslint-plugin-import and eslint.
- Update node modules: async, lodash, numbro and request.
- Update node modules: lodash, eslint-plugin-import, eslint-config-airbnb and eslint.
- Fix eslint check issues.
- Update node modules: lodash, gulp-imagemin, eslint-plugin-import, eslint-config-airbnb and eslint.
- Update numbro, bluebird, cheerio and eslint.
- Update bluebird, lodash, moment and request.

### v0.3.8

- Update bluebird and lodash.
- Fix help content output for node.js V6+;
- Update node.js modules: async, eslint, eslint-plugin-import, gulp-imagemin and js-yaml.

### v0.3.7

- Fix SZ market insider trading query for the DOM structure has been changed.

### v0.3.6

- Add insider trading query support for quotes start with `001`.
- Fix finance calendar: remove redundant spaces, etc.
- Add quotes begin with `001` to SZ market.
- Update node modules:lodash,request and blessed-contrib.
- Fix eslint checking warnings.
- Update node modules:bluebird, js-yaml and moment.
- Update node modules:blessed-contrib,bluebird,imagemin-pngquant and lodash.
- Update node modules:cheerio,js-yaml,lodash,moment,request and bluebird.

### v0.3.5

- Fix javascript coding style.
- Update lodash to v4.0.0: change _.pluck,_.trimRight,_.trimLeft,_.padRight, etc.
- Update node modules:async, js-yaml and moment.
- Use columnify to replace printf.
- Remove grunt tasks.
- Update node modules:cheerio,js-yaml,lodash,moment,request and bluebird.
- Fix eslint check warnings.
- Update node modules:blessed-contrib,bluebird,imagemin-pngquant and lodash.
- Update node modules:bluebird, js-yaml and moment.
- Update node modules:lodash,request and blessed-contrib.
- Update node modules:lodash,async.
- Add quotes begin with `001` to SZ market.

### v0.3.3

- Update node modules:bluebird, commander, js-yaml, blessed-contrib and request.
- Add gulp building tasks.
- Update async to v1.5.0.
- Update blessed-contrib to v2.5.1.
- Update node modules:blessed-contrib,js-yaml,request.
- Use 'SystemDrive' instead of 'USERPROFILE' in `process.env` for windows users to store conf file.
- Use template string of ES6 instead of string concat, node.js v4.0.0 or greater required.
- Update node modules:async, blessed-contrib and moment.


### v0.3.2

- Add float number support for star.
- Update node.js module: js-yaml.
- Fix #41:Add filter margin symbols support.
- Update node modules:bluebird, iconv and request.

### v0.3.1

- Fix pagination bugs for insider trading query with specified security code.
- Update node module request to 2.61.0.
- Add grep from security code feature in stock trace.

### v0.3.0

- Fix Insider trading query paging bug.
- Update node.js modules:blessed@0.1.20,blessed-contrib@2.3.3,grunt-eslint@17.0.0.
- Update node.js module:lodash@3.10.1.
- Add insider trading query for none specific stocks.
- Add `--market` support while querying insider tradings:SZM-深圳主板, SZGEM-深圳创业板, SZSME-深圳中小板, SHM-上海主板.
- Add `-i --top-buy` param to query top buy value of insider tradings, time span support: 1m~12m.
- Add `-i --top-sell` param to query top sell value of insider tradings, time span support: 1m~12m.
- Add grunt task for image optimization.
- Update node.js module:async,blessed,grunt-eslint,iconv.

### v0.2.8

- Add sort by `bdiff` feature to sort stocks by `(s.price - s.cheap)/s.price`.
- Add sort by `sdiff` feature to sort stocks by `(s.price - s.expensive)/s.price`.
- Add `--remove` param to remove the symbols from result with the specified keywords in name or comment.
- Update node modules: moment to 2.10.6.

### v0.2.7

- Add `--lteb [pct]` support to query symbols that make 100*(s.price - s.cheap)/s.price <= pct.
- Add `--gtes [pct]` support to query symbols that make 100*(s.price - s.expensive)/s.price >= pct.
- Update node modules:blessed,blessed-contrib,moment,request.

### v0.2.6

- Make `star -wo` or `star -w -o` to watch held stocks.
- Use moment 'zh-cn' locale instead of const days to format dayOfWeek output in finance cal.
- Update node.js modules:async@~1.4.0,blessed@~0.1.14,blessed-contrib@~2.3.1.
- Add `--lteb` and `--gtes` support to filter the stocks whose current price is lower than buy price or greater than sell price.
- Fix bug of query none exist symbols.

### v0.2.5

- Fix Trailing comma bugs when run `star *` cmds.
- Set conf.chunkSize from 20 to 25.
- Improve console output string padding problem.
- Add support for query latest insider tradings.
- Add latest insider trading summary limit.
- Update request timeout error msg for SZ market insider trading query.


### v0.2.4

- Update blessed-contrib to v2.2.1
- Trim inline tabs of events in finance calendar.
- Update bluebird to v2.9.34.
- Add trace symbol duplication check.


### v0.2.3

- Add finance calendar feature.
- Add day of week field for finance calendar.


### v0.2.2

- Publish star to npm and cnpm.


### v0.2.0

- Update node modules: blessed@0.1.7/lodash@3.10.0/bluebird@2.9.32/grunt-eslint@16.0.0
- Add insider trading query feature.
- Add `--from`, `--to`, `--span` param for insider trading query.

star.js
#!/usr/bin/env node

/**
 * @author      hustcer
 * @license     MIT
 * @copyright   uzfin.com
 * @create      05/18/2015
 */

'use strict';

let _       = require('lodash'),
    colors  = require('colors'),
    Promise = require('bluebird'),
    cmd     = require('commander');

const pkg   = require('./package.json');
let   conf  = require('./lib/conf').conf;
let action  = null;   // cmd action

/**
 * Available colors are:
 * bold, italic, underline, inverse, yellow, cyan, white,
 * green, red, grey, blue, rainbow, zebra, random, magenta,
 */
const COLOR_THEME = {
    error  : 'red',
    info   : 'blue',
    data   : 'grey',
    header : 'blue',
    warn   : 'yellow',
    em     : 'magenta',
};

colors.setTheme(COLOR_THEME);

// Fix Issue: https://github.com/tj/commander.js/issues/530
// For Node v6.0.0 and v6.1.0; TODO: remove it later
[process.stdout, process.stderr].forEach(stream => {
  // eslint-disable-next-line
  if (stream._handle && typeof stream._handle.setBlocking === 'function') { stream._handle.setBlocking(true); }
});

cmd
  .version(pkg.version)
  .usage(`[options]${' OR'.em} star code1,code2，code3...，codeN`)
  .description('Star is a command line tool for STock Analysis and Research.')
  .option('-a, --all'          , 'display all stocks.')
  .option('-o, --hold'         , 'display all held stocks.')
  .option('-M, --margin'       , 'display stocks that support Margin.')
  .option('-C, --cal'          , 'display finance calendar of the future month.')
  .option('-I, --ignore'       , 'display all ignored stocks.')
  .option('-i, --insider [c]'  , 'display insider trading records of specified stocks. data source: sse or szse')
  .option('    --code [code]'  , 'specify the stock code of insider tradings you want to query. data source: uzfin.com')
  .option('    --market <mkt>' , (`specify the market of insider trading query, case insensitive:SZM-深圳主板, SZGEM-深圳创业板, SZSME-深圳中小板,
                     SHM-上海主板. multiple market should be separated by "," or "，". `))
  .option('    --top-buy'      , 'query top buy of insider tradings, should be used with "-i" or "--insider". time span:1m~12m.')
  .option('    --top-sell'     , 'query top sell of insider tradings, should be used with "-i" or "--insider". time span:1m~12m.')
  .option('    --latest-sz'    , 'query latest insider tradings of ShenZhen market, should be used with "-i" or "--insider". data source: szse')
  .option('    --latest-sh'    , 'query latest insider tradings of ShangHai market, should be used with "-i" or "--insider". data source: sse')
  .option('    --show-detail'  , 'show detail latest insider trading records, should be used with "-i" or "--insider".')
  .option('-w, --watch [c1...]', 'watch specified stocks or watch all the stocks in watch list.')
  .option('-r, --reverse'      , 'sort stocks in ascending order according to designated field.')
  .option('-l, --limit <n>'    , 'set total display limit of current page.', parseInt)
  .option('-p, --page  <n>'    , 'specify the page index to display.', parseInt)
  .option('-d, --data  <data>' , 'specify data provider, should be one of "sina" or "tencent".')
  .option('-f, --file  <file>' , 'specify the symbol file path.')
  .option('--from <2014/06/01>', 'specify the beginning date of insider tradings.')
  .option('--to <2015/07/09>'  , 'specify the ending date of insider tradings.')
  .option('--span <3m>'        , 'specify the month span of insider tradings ahead of today, should be 1m~24m.')
  .option('-L, --lte   <pct> ' , 'filter the symbols whose upside potential is lower than or equal to the specified percentage.', parseInt)
  .option('-G, --gte   <pct> ' , 'filter the symbols whose upside potential is greater than or equal to the specified percentage.', parseInt)
  .option('-U, --under <star>' , 'filter the symbols whose star is under or equal to the specified star value.', parseInt)
  .option('-A, --above <star>' , 'filter the symbols whose star is above or equal to the specified star value.', parseInt)
  .option('--lteb [pct]'       , "filter the symbols whose current price is lower than or equal to it's buy/cheap price", parseInt)
  .option('--gtes [pct]'       , "filter the symbols whose current price is greater than or equal to it's sell/expensive price", parseInt)
  .option('-g, --grep  <kw>  ' , 'specify the keyword to grep in name or comment, multiple keywords should be separated by ",".')
  .option('--remove  <kw>    ' , (`remove the symbols from result with the specified keywords in name or comment, multiple keywords
                     should be separated by ",".`))
  .option('-e, --exclude <pre>', (`exclude stocks whose code number begin with: 300,600,002 or 000, etc. multiple prefixs can be
                     used and should be separated by "," or "，". `))
  .option('-c, --contain <pre>', `display stocks whose code number begin with: 300,600,002 or 000, etc. multiple prefixs can be
                     used and should be separated by "," or "，". `)
  .option('-s, --sort <sort>'  , (`specify sorting field, could be sorted by: code/star/price/targetp/incp/bdiff/sdiff/capacity/
                     pe/pb to sort stocks by code, star, price, price to target percent, price increase percent,
                     (s.price - s.cheap)/s.price, (s.price - s.expensive)/s.price, capacity, PE and PB separately.
                     and sort by capacity/pe/pb only works while using tencent data source.`))
  .parse(process.argv);

let actions = {
    WATCH: function watch(){

        let Watch = require('./lib/watch').Watch;
        Watch.doWatch(cmd.watch);

    },
    CAL: function cal(){

        let Cal = require('./lib/cal').Cal;
        Cal.showCal();

    },
    INSIDER: function insider(){

        let async   = require('async');
        let Insider = require('./lib/insider').Insider;

        if(cmd.latestSz){ Insider.querySZLatest();    return false; }
        if(cmd.latestSh){ Insider.querySHLatest();    return false; }
        if(cmd.topBuy)  { Insider.queryTopList('BV'); return false; }
        if(cmd.topSell) { Insider.queryTopList('SV'); return false; }
        if(cmd.insider === true){ Insider.queryMiscInsider(); return false; }

        let query   = cmd.insider.replace(/，/g, ',');
        let symbols = _.trimEnd(query, ',').split(',');
        if(symbols.length > conf.chunkSize){
            console.error((`You can query at most ${conf.chunkSize} symbols once.`).error);
            return false;
        }

        async.eachSeries(symbols, (c, callback) => {
            Insider.queryInsider(c, callback);
        }, () => {
            console.log('ALL DONE!');
        });
        return false;

    },
    QUERY: function query(){

        let Query = require('./lib/query').Query;
        Query.doQuery(cmd.args[0]);

    },
    TRACE: function trace(){

        let Trace   = require('./lib/trace').Trace;
        let symbols = Trace.getFilteredSymbols();

        if(!symbols){ return false; }

        let symList = _.chunk(symbols, conf.chunkSize);

        Promise
            .resolve(symList)
            .each(syms => Trace.querySymbols(syms))
            .then(()   => Trace.printResults())
            .then(()   => Trace.printSummary());

        return false;
    },
};

let doCmd = function doCmd() {

    // 'TRACE' is the default action
    action  = 'TRACE';

    if(cmd.watch)  {  action = 'WATCH';    }
    if(cmd.insider){  action = 'INSIDER';  }
    if(cmd.cal)    {  action = 'CAL';      }

    if(cmd.args.length === 1 ){

      action = 'QUERY';

    } else if (cmd.args.length > 1) {

      console.error('Input error, please try again, or run "star -h" for more help.'.error);
      return false;
    }

    actions[action]();
    return false;
};

// Get the Job Done!
doCmd();

